File: packages/java/ygo74-agent-runtime/src/main/java/com/ygo74/agentruntime/middleware/MiddlewarePipeline.java
package com.ygo74.agentruntime.middleware;

import com.ygo74.agentruntime.domains.contracts.ExchangeModels.StandardExchangeResponse;

import java.util.List;
import java.util.function.Function;

public final class MiddlewarePipeline {
    private MiddlewarePipeline() {}

    public static StandardExchangeResponse execute(
            Middleware.MessagePipelineContext context,
            List<Middleware> middlewares,
            Function<Middleware.MessagePipelineContext, StandardExchangeResponse> handler) {
        return chain(0, context, middlewares, handler);
    }

    private static StandardExchangeResponse chain(
            int index,
            Middleware.MessagePipelineContext context,
            List<Middleware> middlewares,
            Function<Middleware.MessagePipelineContext, StandardExchangeResponse> handler) {
        if (index >= middlewares.size()) {
            return handler.apply(context);
        }

        return middlewares.get(index).invoke(context, next -> chain(index + 1, next, middlewares, handler));
    }
}

File: packages/dotnet/Ygo74.AgentRuntime/Middleware/MiddlewarePipeline.cs
using Ygo74.AgentRuntime.Domains.Contracts;

namespace Ygo74.AgentRuntime.Middleware;

public static class MiddlewarePipeline
{
    public static StandardExchangeResponse Execute(
        MessagePipelineContext context,
        IReadOnlyList<IMiddleware> middlewares,
        Func<MessagePipelineContext, StandardExchangeResponse> handler)
    {
        StandardExchangeResponse Next(int index, MessagePipelineContext ctx)
        {
            if (index >= middlewares.Count)
            {
                return handler(ctx);
            }

            return middlewares[index].Invoke(ctx, next => Next(index + 1, next));
        }

        return Next(0, context);
    }
}
